### PR TITLE
Fix DataFrame.agg produces different types if the DataFrame is empty

### LINKED ISSUE #41672: BUG: DataFrame.agg produces different types if the DataFrame is empty
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
tdf2 = pd.DataFrame([], columns=['lang', 'name'])
print(type(tdf2.agg({'name': lambda y: y.values})))

tdf2 = pd.DataFrame([['a', 'boof']], columns=['lang', 'name'])
print(type(tdf2.agg({'name': lambda y: y.values})))
```

#### Problem description

This seems very similar to the description of #16621, and may be the same bug; check there for some history, which points to an original fix in #2476. You can also look at [this StackExchange question](https://stackoverflow.com/questions/67697606/pandas-dataframe-agg-produces-multiindex-after-selection-with-missing-categories/) if you want to see how I nearly pulled my hair out.

The first call of `.agg()` above, on the empty DataFrame, produces a DataFrame result. The second call of `.agg()` produces a Series instead.

This has unfortunate implications when used with `groupby()`:

```python
tdf2 = (
    pd.DataFrame([['a', 'boof'], ['a', 'toop']], columns=['lang', 'name'])
    .astype({'lang': pd.CategoricalDtype(['a', 'b'])})
)
print(tdf2.groupby('lang').apply(lambda x: x.agg({'name': lambda y: ', '.join(y.values)})))
```

...produces:

| lang | | 0 | name |
| -- | -- | -- | -- |
| **a** | **name** | boof, toop | NaN |

...instead of:

| lang | name |
| -- | -- |
| **a** | boof, toop |

#### Expected Output

I'm assuming that the call with data produces a series because the `agg()` is single-valued, therefore the top code should probably produce two Series:

```
<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>
```

#### Actual output

```
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
```

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit           : 2cb96529396d93b46abab7bbc73a208e708c642e
python           : 3.8.8.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Mon Apr 12 20:57:45 PDT 2021; root:xnu-6153.141.28.1~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.4
numpy            : 1.20.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 52.0.0.post20210125
Cython           : 0.29.23
pytest           : 6.2.3
hypothesis       : None
sphinx           : 4.0.1
blosc            : None
feather          : None
xlsxwriter       : 1.3.8
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.3
IPython          : 7.22.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.9.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.4
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.7
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1
```

</details>

### PR DESCRIPTION
**Pull request by weikhor**
- [ ✔] closes https://github.com/pandas-dev/pandas/issues/41672
- [ ❌] tests added / passed
- [ ❌] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

File: pandas/tests/frame/test_aggregate.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+from pandas import (
+    DataFrame,
+    Series,
+)
+
+import pandas._testing as tm
+
+
+def _check_mixed_int(df, dtype=None):
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_aggregate.py:13]: is this supposed to be a test?

Comment (weikhor) [reply pandas/tests/frame/test_aggregate.py]: My mistake. Have updated the code

File: pandas/tests/frame/test_aggregate.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+
```
Comment (jreback) [inline pandas/tests/frame/test_aggregate.py:2]: these go in pandas/tests/apply/test_frame_apply.py

there are already extensive tests, so find a good place to exercise the empty case

we have these, so model after

```
(pandas) jreback@lhs2:~/pandas-dev$ grep -r empty pandas/tests/apply/|grep def
pandas/tests/apply/test_frame_transform.py:def test_transform_empty_listlike(float_frame, ops, frame_or_series):
pandas/tests/apply/test_frame_transform.py:def test_transform_empty_dictlike(float_frame, ops, frame_or_series):
pandas/tests/apply/test_frame_transform.py:def test_transform_empty_dataframe():
pandas/tests/apply/test_series_apply.py:def test_apply_empty_integer_series_with_datetime_index():
pandas/tests/apply/test_series_apply.py:def test_map_empty(index):
pandas/tests/apply/test_frame_apply.py:def test_apply_empty(float_frame):
pandas/tests/apply/test_frame_apply.py:def test_apply_with_reduce_empty():
pandas/tests/apply/test_frame_apply.py:def test_apply_funcs_over_empty(func):
pandas/tests/apply/test_frame_apply.py:def test_nunique_empty():
pandas/tests/apply/test_frame_apply.py:def test_apply_empty_infer_type():
pandas/tests/apply/test_frame_apply.py:def test_apply_empty_list_reduce():
```

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
Thanks @weikhor for the PR. can you add a release note to doc/source/whatsnew/v1.3.1.rst, Bug fixes section and address https://github.com/pandas-dev/pandas/pull/41722#discussion_r642685684

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears this has gone stale and needs addressing of the reviews. Closing, but happy to reopen if still interested.

**Pull request closed without merging**